pipelines/controller/src/main/java/org/openmrs/analytics/PipelineManager.java

```diff
@@ -137,23 +142,69 @@ private void initDwhStatus() {
           logger.debug("Found a more recent DWH {}", path.getFilename());
           lastDwh = path.getFilename();
         }
+        // Do not consider if the DWH is not completely created earlier.
+        if (!dwhFilesManager.isDwhComplete(path)) {
+          continue;
+        }
+        if (lastCompletedDwh.isEmpty() || lastCompletedDwh.compareTo(path.getFilename()) < 0) {
+          logger.debug("Found a more recent completed DWH {}", path.getFilename());
+          lastCompletedDwh = path.getFilename();
+        }
       }
     } catch (IOException e) {
       throw new RuntimeException(e);
     }
-    if (lastDwh.isEmpty()) {
+    if (lastCompletedDwh.isEmpty()) {
       logger.info("No DWH found; it should be created by running a full pipeline");
       currentDwh = null;
       lastRunEnd = null;
     } else {
-      logger.info("Initializing with most recent DWH {}", lastDwh);
+      logger.info("Initializing with most recent DWH {}", lastCompletedDwh);
       ResourceId resourceId =
           FileSystems.matchNewResource(baseDir, true)
-              .resolve(lastDwh, StandardResolveOptions.RESOLVE_DIRECTORY);
+              .resolve(lastCompletedDwh, StandardResolveOptions.RESOLVE_DIRECTORY);
       currentDwh = DwhFiles.forRoot(resourceId.toString());
       // There exists a DWH from before, so we set the scheduler to continue updating the DWH.
       lastRunEnd = LocalDateTime.now();
     }
+    if (!lastDwh.isEmpty()) {
+      initialiseLastRunDetails(baseDir, lastDwh);
+    }
+  }
+
+  /**
+   * This method initialises the lastRunDetails based on the dwh snapshot created recently. In case
+   * an incremental run exists in the snapshot, the status is set based on the incremental run.
+   */
+  private void initialiseLastRunDetails(String baseDir, String dwhDirectory) {
+    try {
+      List<ResourceId> paths =
+          dwhFilesManager.getAllChildDirectories(baseDir + "/" + dwhDirectory).stream()
+              .collect(Collectors.toList());
+      for (ResourceId path : paths) {
```

Comment:
Instead of fetching all paths then check for the incremental one with a loop, can't we directly get the incremental one and avoid the loop?

Reply:
Yes, this is a good point, I have made the changes accordingly in the latest commit.